core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -424,79 +457,50 @@ protected boolean checkBlocks() {
                 }
                 return false;
             }
+            // request does not have a shardId yet, we need to pass the concrete index to resolve shardId
+            resolveRequest(clusterState, concreteIndex, request);
+            assert request.shardId() != null : "request shardID must be set in resolveRequest";
             return true;
         }
 
-        protected ShardRouting resolvePrimary(ShardIterator shardIt) {
-            // no shardIt, might be in the case between index gateway recovery and shardIt initialization
-            ShardRouting shard;
-            while ((shard = shardIt.nextOrNull()) != null) {
-                // we only deal with primary shardIt here...
-                if (shard.primary()) {
-                    return shard;
-                }
-            }
-            return null;
-        }
-
-        /**
-         * send the request to the node holding the primary or execute if local
-         */
-        protected void routeRequestOrPerformLocally(final ShardRouting primary, final ShardIterator shardsIt) {
-            if (primary.currentNodeId().equals(observer.observedState().nodes().localNodeId())) {
-                try {
-                    threadPool.executor(executor).execute(new AbstractRunnable() {
-                        @Override
-                        public void onFailure(Throwable t) {
-                            finishAsFailed(t);
-                        }
+        private void performAction(final String nodeId, final String action) {
+            DiscoveryNode node = observer.observedState().nodes().get(nodeId);
+            final boolean isPrimaryAction = action.equals(transportPrimaryAction);
```

Comment:
Nit: I think it will be safer to have this boolean as a parameter and determine the action here. I'm weary of arbitrary string input.